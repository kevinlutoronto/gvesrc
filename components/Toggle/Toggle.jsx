// @flow
import React from 'react';
import cx from 'classnames';
import _ from 'lodash';
import { getDataAttrs } from '../../dataUtils';
import type { Data } from '../../types';
import style from './style.scss';

const type = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical',
  SMALL: 'small',
};

type Props = {|
  items: Array<{
    label: string,
    title: string,
    disabled?: boolean,
  }>,
  onChange: (string, number) => void,
  selectedIndex?: number,
  selectedLabel: string,
  type: string,
  data?: Data,
  size?: string,
  className?: string,
|};

class Toggle extends React.Component<Props> {
  static defaultProps = {
    items: [],
    type: type.HORIZONTAL,
    onChange: () => {},
  };

  onClick(item: Object, index: number) {
    if (!item.disabled) {
      this.props.onChange(item.label, index);
    }
  }

  onKeyDown(event: Object, item: Object, index: number) {
    if (event.keyCode === 13 /* enter */) {
      this.onClick(item, index);
    }
  }

  render() {
    return (
      <ul {...getDataAttrs(this.props.data)} className={style.toggleWrapper} role="tablist">
        {_.map(this.props.items, (item, index) => {
          const isActive =
            this.props.selectedIndex === index || this.props.selectedLabel === item.label;
          const toggleClasses = cx(style.toggle, this.props.className, {
            [style.active]: isActive,
            [style.disabled]: item.disabled,
            [style.horizontal]: this.props.type === type.HORIZONTAL,
            [style.vertical]: this.props.type === type.VERTICAL,
            [style.small]: this.props.size === type.SMALL,
          });

          return (
            <li
              role="tab"
              onClick={() => this.onClick(item, index)}
              onKeyDown={event => this.onKeyDown(event, item, index)}
              className={toggleClasses}
              key={index}
              type={this.props.type}
              data-state={item.disabled ? 'disabled' : 'enabled'}
              data-active={isActive}
            >
              {item.title}
            </li>
          );
        })}
      </ul>
    );
  }
}

Toggle.displayName = 'Plasma@Toggle';

export default Toggle;
