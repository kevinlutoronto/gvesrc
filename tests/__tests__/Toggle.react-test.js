/* eslint-disable */

import React from 'react';
import Toggle from '../../components/Toggle/Toggle';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import uuidv4 from 'uuid/v4';

describe('Toggle Component', () => {
  test('render', () => {
    const component = renderer.create(<Toggle />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('two items', () => {
    const items = [
      {
        label: 'one',
        title: 'One',
      },
      {
        label: 'two',
        title: 'Two',
      },
    ];
    const component = shallow(<Toggle items={items} />);
    const options = component.find('li');
    expect(options).toHaveLength(2);
    options.forEach(option => {
      expect(option.prop('data-state')).toEqual('enabled');
    });
  });

  describe('clicks', () => {
    const items = [
      {
        label: 'one',
        title: 'One',
      },
      {
        label: 'two',
        title: 'Two',
        disabled: true,
      },
    ];
    const onChange = jest.fn();
    const component = mount(<Toggle items={items} onChange={onChange} />);

    beforeEach(() => {
      onChange.mockClear();
    });

    test('enabled option is clickable', () => {
      const enabledOptions = component.find('li[data-state="enabled"]');
      expect(enabledOptions).toHaveLength(1);

      const enabledOption = enabledOptions.first();
      expect(enabledOption.text()).toEqual(items[0].title);

      enabledOption.simulate('click');
      expect(onChange).toHaveBeenCalledWith(items[0].label, 0);
    });

    test('enabled option accepts <enter> keydown', () => {
      const enabledOption = component.find('li[data-state="enabled"]').first();
      enabledOption.simulate('keydown', { keyCode: 13 });
      expect(onChange).toHaveBeenCalledWith(items[0].label, 0);
    });

    test('disabled options are not clickable', () => {
      const disabledOptions = component.find('li[data-state="disabled"]');
      expect(disabledOptions).toHaveLength(1);
      const disabledOption = disabledOptions.first();
      expect(disabledOption.text()).toEqual(items[1].title);
      disabledOption.simulate('click');
      disabledOption.simulate('keydown', { keyCode: 13 });
      expect(onChange).toHaveBeenCalledTimes(0);
    });
  });

  test('data attributes are added', () => {
    const items = [
      {
        label: 'one',
        title: 'One',
      },
    ];
    const dataAttributes = {
      qa: 'toggle',
      id: uuidv4(),
    };
    const component = shallow(<Toggle items={items} data={dataAttributes} />);

    Object.entries(dataAttributes).forEach(([key, value]) =>
      expect(component.prop(`data-${key}`)).toEqual(value)
    );
  });
});
