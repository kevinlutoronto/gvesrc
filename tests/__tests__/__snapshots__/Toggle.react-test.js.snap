// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Component render 1`] = `
<ul
  className="toggleWrapper"
  role="tablist"
/>
`;
